framework/meta/src/meta_cli.rs

```diff
@@ -0,0 +1,83 @@
+use super::{meta_config::MetaConfig, output_contract::OutputContractConfig};
+use crate::{
+    cli_args::{CliAction, CliArgs},
+    sc_upgrade::upgrade_sc,
+};
+use clap::Parser;
+use multiversx_sc::{
+    abi::ContractAbi, api::uncallable::UncallableApi, contract_base::ContractAbiProvider,
+};
+
+/// The ABI provider set when
+struct NoAbiProvider;
+
+impl NoAbiProvider {
+    pub const NAME: &str = "no-abi";
+}
+
+impl ContractAbiProvider for NoAbiProvider {
+    type Api = UncallableApi;
+
+    fn abi() -> ContractAbi {
+        ContractAbi {
+            name: Self::NAME,
+            ..Default::default()
+        }
+    }
+}
+
+fn process_abi<AbiObj: ContractAbiProvider>(cli_args: &CliArgs) -> Option<MetaConfig> {
+    let input_abi = <AbiObj as ContractAbiProvider>::abi();
+    if input_abi.name == NoAbiProvider::NAME {
+        return None;
+    }
+
+    super::meta_validate_abi::validate_abi(&input_abi).unwrap();
+    let mut meta_config = MetaConfig::create(input_abi, cli_args.load_abi_git_version);
+    meta_config.write_abi();
+    meta_config.generate_wasm_crates();
+    Some(meta_config)
+}
+
+pub fn cli_main_standalone() {
+    cli_main::<NoAbiProvider>();
+}
+
+pub fn cli_main<AbiObj: ContractAbiProvider>() {
+    let cli_args = CliArgs::parse();
+
+    let meta_config_opt = process_abi::<AbiObj>(&cli_args);
+
+    if let Some(command) = cli_args.command {
```

Comment:
I feel like this can be rewritten to look slightly better. Normally, I'm also a fan of the `let Some() = variable` patter, but not when it leads to very nested code like this.

I suggest something like:
```
if cli_args.command.is_none() { return; }

let command = unsafe { cli_args.command.unwrap_unchecked() };
if meta_config_opt.is_none()  {
       match command {
CliAction::Upgrade(args) => {
                upgrade_sc(&args);
            },
          _ => // error

        return;
}

let meta_config = unsafe { meta_config_opt.unwrap_unchecked() };
// your match here, but without any "expect"
```

Sorry for bad formatting, blame github. Also, if you want a better formatted message, you can have a method for `CliAction` which gives you the command name as string, i.e. `build`, `build-dbg`, etc.

Reply:
I split the CLI args in two and removed the Option. It is cleaner now.